pkg/auth/authorizer/abac/abac.go

```diff
@@ -79,62 +64,150 @@ func NewFromFile(path string) (policyList, error) {
+				return nil, policyLoadError{path, i, b, err}
+			}
+		} else {
+			if err := latest.Codec.DecodeInto(b, p); err != nil {
+				return nil, policyLoadError{path, i, b, err}
+			}
 		}
+
 		pl = append(pl, p)
 	}
 
+	if unversionedLines > 0 {
+		glog.Warningf(`Policy file %s contained unversioned rules. See docs/admin/authorization.md#abac-mode for ABAC file format details.`, path)
+	}
+
 	if err := scanner.Err(); err != nil {
-		return nil, err
+		return nil, policyLoadError{path, -1, nil, err}
 	}
 	return pl, nil
 }
 
-func (p policy) matches(a authorizer.Attributes) bool {
-	if p.subjectMatches(a) {
-		if p.Readonly == false || (p.Readonly == a.IsReadOnly()) {
-			if p.Resource == "" || (p.Resource == a.GetResource()) {
-				if p.Namespace == "" || (p.Namespace == a.GetNamespace()) {
-					return true
-				}
-			}
-		}
+func matches(p api.Policy, a authorizer.Attributes) bool {
+	if !subjectMatches(p, a) {
+		return false
 	}
-	return false
+	if !verbMatches(p, a) {
+		return false
+	}
+	return resourceMatches(p, a) || nonResourceMatches(p, a)
 }
 
-func (p policy) subjectMatches(a authorizer.Attributes) bool {
-	if p.User != "" {
-		// Require user match
-		if p.User != a.GetUserName() {
+// subjectMatches returns true if specified user and group properties in the policy match the attributes
+func subjectMatches(p api.Policy, a authorizer.Attributes) bool {
+	matched := false
+
+	switch p.User {
+	case "":
+		// not specified
+	case "*":
+		matched = true
+	default:
+		matched = p.User == a.GetUserName()
+		if !matched {
 			return false
 		}
 	}
 
-	if p.Group != "" {
-		// Require group match
+	switch p.Group {
+	case "":
+		// not specified
+	case "*":
+		matched = true
+	default:
+		groupMatched := false
 		for _, group := range a.GetGroups() {
 			if p.Group == group {
-				return true
+				groupMatched = true
+				matched = true
 			}
 		}
+		if !groupMatched {
+			return false
+		}
+	}
+
+	return matched
+}
+
+func verbMatches(p api.Policy, a authorizer.Attributes) bool {
+	// Reject if policy only allows readonly access and this is not a readonly request
+	if p.Readonly && !a.IsReadOnly() {
 		return false
 	}
+	// TODO: match on verb
+	return true
+}
 
+func nonResourceMatches(p api.Policy, a authorizer.Attributes) bool {
+	// A non-resource policy cannot match a resource request
+	if a.IsResourceRequest() {
+		return false
+	}
+	// Allow wildcard match
+	if p.NonResourcePath == "*" {
+		return true
+	}
+	// Allow exact match
+	if p.NonResourcePath == a.GetPath() {
+		return true
+	}
+	// Allow a trailing * subpath match
+	if strings.HasSuffix(p.NonResourcePath, "*") && strings.HasPrefix(a.GetPath(), strings.TrimRight(p.NonResourcePath, "*")) {
+		return true
+	}
+	return false
+}
+
+func resourceMatches(p api.Policy, a authorizer.Attributes) bool {
+	// A resource policy cannot match a non-resource request
+	if !a.IsResourceRequest() {
+		return false
+	}
+	if p.Namespace != "*" && p.Namespace != a.GetNamespace() {
```

Comment:
> returning true requires nesting... false lets us exit early... I can swap if you want

I think of authorization evaluation in terms of successively tighter checks, so nesting makes more sense to me.  "Match this and this and this and this".  Having seen both, you prefer this?